### PR TITLE
Dynamic URLs

### PR DESCRIPTION
**Pull request by kaldis-berzins**
By submitting this PR, I am indicating to the Numberscope maintainers that I have read and understood the [contributing guidelines](https://numberscope.colorado.edu/doc/CONTRIBUTING/) and that this PR follows those guidelines to the best of my knowledge. I have also read the [pull request checklist](https://numberscope.colorado.edu/doc/doc/pull-request-checklist/) and followed the instructions therein.

<hr/>
This PR is first in the PR merge chain.

# URLs
## Specimen
 - Changed the constructor to automatically pull from the export modules instead of requiring an already constructed visualizer and sequence to be passed in.
 - Changes getters and setters to reflect that keys are the primary way to interface with the specimen now.
 - Adds a fromURL function that constructs a specimen from a url.
## Scope
  - Updates URL when changes are made to a sequence/visualizer.
## Sequences
 - Are either a family (constructed for each specimen) or an instance (singleton for all)
 - Each export module contains a default name and description dedicated for the gallery and modal switchers (this applies as well to visualizers)

### CODE REVIEW

**Comment by gwhitney** [comment]
Thanks for submitting these PRs and clearly marking the desired merge order. I will start by reviewing this one. If we don't get through the usual back-and-forth of review for the end of the semester, we will take it from whatever point we do reach.

But one initial question about this PR: a good fraction of the changes consist of changing the export modules of sequences and visualizers to get their names and descriptions from constants in their source files, rather than from properties of the classes that define them, roughly reverting a change from the prior PR. It would help us to understand:
(a) What is the motivation for these changes?
(b) Are they needed for the URL encoding/decoding of Specimen parameters, or is it a separate change that happened to be included in this PR?

**Comment by gwhitney** [comment]
> (a) What is the motivation for these changes?

Ah, I think I understand now. I think it is because the approach from the previous PR actually simply didn't work, even though it typechecked and compiled and seemed to run. I will continue the review on that basis, although I am unfortunately out of time for tonight.

File: src/shared/Specimen.ts [hunk 1]
```diff
@@ -52,23 +87,53 @@ export class Specimen {
...
+     */
+    set visualizer(visualizerKey: string) {
+        this.visualizerKey = visualizerKey
+    }
+    /**
+     * Exists for redundancy, same as set sequenceKey()
+     */
+    set sequence(sequenceKey: string) {
+        this.sequenceKey = sequenceKey
+    }
```
Comment (gwhitney) [inline src/shared/Specimen.ts:136]: Why do we want redundant names for these two setters? Redundancy in interfaces is in general a bad thing, not a good thing. Please remove unless there's a strong argument for having these setters in addition to visualizerKey and sequenceKey.

File: src/visualizers/Turtle.ts [hunk 2]
```diff
@@ -126,8 +129,8 @@ class Turtle extends P5Visualizer(paramDesc) {
         super.setup()
         this.currentIndex = this.seq.first
         this.orientation = 0
-        this.X = this.sketch.width / 2 + this.start.x
-        this.Y = this.sketch.height / 2 + this.start.y
+        this.X = this.sketch.width / 2
+        this.Y = this.sketch.height / 2
```
Comment (gwhitney) [inline src/visualizers/Turtle.ts:134]: These changes to this.X and this.Y are a merge error, reverting a change from the prior PR. Please restore the lines as they are in ui2 now, thanks.

**Review by gwhitney — COMMENTED** [review]
OK, I have done a code review. I still need to manually test the behavior. In the meantime, @katestange and/or @Vectornaut please could you look over the review comments and say whether you think these are all changes that we could make in an additional commit we would make to this PR, or whether we prefer the Delft team to respond if they have a chance before the semester is over? Of course, if anyone on the Delft team has time simply to respond to these comments and make appropriate changes, they should feel free.

**Comment by gwhitney** [comment]
OK, I have found the URLs to update and load correctly across some different Visualizers and Sequences. I could not "trick" the system into producing a bad URL by e.g. entering a formula that would not parse and then copying the URL. This seems to work well and be reasonably limited in scope. So I am comfortable with merging it as soon as the above code review comments are addressed. I am willing to add a commit or two that addresses them myself in the interest of expediency, if that seems reasonable to either @katestange or @Vectornaut, so please just let me know (or the team can just go ahead and address them if someone has time). Thanks!

**Comment by katestange** [comment]
> OK, I have found the URLs to update and load correctly across some different Visualizers and Sequences. I could not "trick" the system into producing a bad URL by e.g. entering a formula that would not parse and then copying the URL. This seems to work well and be reasonably limited in scope. So I am comfortable with merging it as soon as the above code review comments are addressed. I am willing to add a commit or two that addresses them myself in the interest of expediency, if that seems reasonable to either @katestange or @Vectornaut, so please just let me know (or the team can just go ahead and address them if someone has time). Thanks!

I would say go for it @gwhitney , that way if something more important to contact the delft team about lies in wait in a later PR, we are more likely to get to it while we still have some contact with them.  That was a complicated sentence, sorry.  Also, you have all the changes fresh in mind and they look fairly simple (merge errors etc).

**Comment by gwhitney** [comment]
OK, in progress, one small commit out of roughly 3 projected.

**Comment by gwhitney** [comment]
OK, have made all of the suggested changes. Merging.

**Pull request merged**